My Store Analytics: Custom Range tab creation UI flow

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11971 

Please do not merge until #11975 is merged first. 

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This adds the following flow:
1. Tapping the "Custom Range" button will open the date range picker,
2. Tapping "Add" in the date range picker will create a new tab called "Custom Range" at the end of the analytics tab (with some dummy content). This new tab should be in selected state.
3. Doing the above will also remove the "Custom Range" button.

Note that this is purely UI flow PR. It does not include:

1. The actual content for the "Custom Range" tab.
2. Persistence (e.g: if the app is restarted, "Custom Range" tab will disappear).

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Run app,
3. Tap the Custom Range button on the analytics tab area (the "calendar +" icon)
4. Ensure date range picker is shown,
5. Play around to make sure date selection functionality is working, then tap "Add",
6. Ensure that you're returned to the analytics area, and a new tab called "Custom Range" is shown.
7. Ensure that the Custom Range button is no longer shown.

## Video
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/266376/f3bb9a96-6171-40f8-a639-cf54f72b4d26

hafizrahman committed:
Add hosting controller for `RangedDatePicker`

hafizrahman committed:
Add navigation coordinator for creating Custom Range tab.

hafizrahman committed:
Connect "Add Custom Range" button with Coordinator.

hafizrahman committed:
Add `onDateRangeSelected` parameter for callback when date range is selected.

hafizrahman committed:
Add a sample stats tab after selecting a date range

hafizrahman committed:
Also remove custom view from superview to truly make it disappear during this function call.

hafizrahman committed:
Improve comment for clarity.

hafizrahman committed:
Simplify function name.

hafizrahman committed:
Add option to customize "Apply" button text in date range picker.

hafizrahman committed:
In custom range stats context, set date range picker apply button text to say "Add".

selanthiraiyan committed:
Update unit test to include 365 in range.

hafizrahman committed:
Use the .custom timeRange as created in #11975

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/CustomRangeTabCreationCoordinator.swift:
```diff
@@ -0,0 +1,52 @@
+        let controller = RangedDatePickerHostingController(
+            datesFormatter: DatesFormatter(),
+            applyButtonTitle: Localization.add,
+            datesSelected: { [weak self] start, end in
```
![Simulator Screen Recording - iPhone 15 Pro Max - 2024-02-13 at 15 10 27](https://github.com/woocommerce/woocommerce-ios/assets/524475/65cbc4ed-6745-48bc-89cf-79061856659c)
Should we add validation to ensure that the start date is less than the end date? Feel free to address this in another PR.

hafizrahman replied:
I agree! This is the existing behavior of the component that we're using. Let's make a new issue to address this.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +436,36 @@ private extension StoreStatsAndTopPerformersViewController {
+    }
+
+    func createCustomRangeTab(_ start: Date, _ end: Date) {
+        print(start.description + end.description)
```
We should remove the `print` statement.

hafizrahman replied:
Nice catch, handled in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/da5fc60b8bf6985ccc501a2a09230135a53c130f

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +436,36 @@ private extension StoreStatsAndTopPerformersViewController {
+                                                                           usageTracksEventEmitter: usageTracksEventEmitter)
+        periodVCs.append(customRangeVC)
+
+        let customRangeTabbedItem = TabbedItem(title: "Custom Range",
```
We should localize the string `"Custom Range"`.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/3b752ee49b6eae5a496dabf0dd8bd2f40b177a8c

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +436,36 @@ private extension StoreStatsAndTopPerformersViewController {
+
+        let customRangeTabbedItem = TabbedItem(title: "Custom Range",
+                                               viewController: customRangeVC,
+                                               accessibilityIdentifier: "todo-identifier")
```
nit: Can we add a TODO with a GH issue to track this?

hafizrahman replied:
This end up not needing an extra todo since I can now set the `accessibilityIdentifier` value (previously was waiting for the enum PR). Handled in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/3b752ee49b6eae5a496dabf0dd8bd2f40b177a8c

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift:
```diff
@@ -102,8 +102,21 @@ class TabbedViewController: UIViewController {
             return
         }
         tabBarStackView.removeArrangedSubview(customTabBarView)
+        customTabBarView.removeFromSuperview()
```
Good catch. 👍

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift:
```diff
@@ -102,8 +102,21 @@ class TabbedViewController: UIViewController {
+        // Setup child view controller
+        items.append(tab)
+        tabBar.items = items
+        configureChildViewControllers()
```
1. I fear that calling this method will add all the viewControllers to the stackView again. (without removing existing views)
2. I also noticed that the custom range view controller is not refreshed/reloaded when added to the tab bar. 

Did you explore the path of rebuilding the `TabbedViewController` again instead of appending a `TabbedItem`. I am asking because I believe that approach will be simpler and less prone to unexpected bugs.

hafizrahman replied:
Thanks for this.

I get the suggestion for point 1. I'll figure out a way to only add the new view controller instead.

Can you explain more about rebuilding the `TabbedViewController`?  I'm asking because `StoreStatsAndTopPerformersViewController`, which has the `createCustomRangeTab()` function, is a subclass of `TabbedViewController`. I'm not sure if you meant we have to rebuild `StoreStatsAndTopPerformersViewController` in order to add the extra tab.


As for this:
> I also noticed that the custom range view controller is not refreshed/reloaded when added to the tab bar.

I intentionally did not take a look at the content displaying behavior yet in this PR, hence why the tab item basically shows and does nothing. I haven't checked how the refresh/reload is being done. My thought is that this should be part of https://github.com/woocommerce/woocommerce-ios/issues/11985

selanthiraiyan replied:
> I intentionally did not take a look at the content displaying behavior yet in this PR, hence why the tab item basically shows and does nothing. I haven't checked how the refresh/reload is being done. My thought is that this should be part of https://github.com/woocommerce/woocommerce-ios/issues/11985

I think it will be better to have an idea about how we are going to handle the lifecycle (load, reload) of this custom tab in this PR. We can handle the actual content building in https://github.com/woocommerce/woocommerce-ios/issues/11985. What do you think? 

I didn't check about how the `StoreStatsAndTopPerformersViewController` works. I suggested recreating `TabbedViewController ` as I thought it would be simpler. It was just a suggestion and if you think we can make the content display and reload behaviour of the custom range work without recreating `StoreStatsAndTopPerformersViewController` please feel free to follow the existing approach. 👍

hafizrahman replied:
I addressed point 1 on https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/c130aa6808b93ad20ad9831cb6d99d8bcd5b33c3 , it simply adds the custom range's view controller to the parent.

hafizrahman replied:
> What do you think?

Yup, I'll handle the lifecycle here.

hafizrahman replied:
I added two relevant commits:

- https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/61f084696817283c675c78a6b15a9a5479710faf This loads the data right after tab creation
- https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/7dd01ae6431772a8094995fed89f604a66d066ef includes the custom tab into the `observeSelectedTimeRangeIndex()`, which does a re-sync on tab change if needed.

Additional note: I observed that in the first commit above, the data loading works. However, on the second commit for the re-sync,  it always goes to `.failure` inside `self.dashboardViewModel.syncSiteVisitStats()`. 

I have not checked deeply yet why this is the case, and I think it should be handled properly in a separate PR to check the backend handling. The commits above are just to ensure that the custom tab is included inside the sync and re-sync process just like the other tabs.

selanthiraiyan replied:
> I have not checked deeply yet why this is the case, and I think it should be handled properly in a separate PR to check the backend handling.

Yes, I agree. Could you please log a sub-task so that we don't miss this?

hafizrahman replied:
Added the task in https://github.com/woocommerce/woocommerce-ios/issues/12014

selanthiraiyan reviewed (commented):
Thanks for working on this, Hafiz! 

I left a few comments. Kindly let me know your thoughts.

hafizrahman committed:
Delete debug code.

hafizrahman committed:
Update customRangeTabbedItem's creation.

hafizrahman committed:
Only add the appended tab's viewController as child.

hafizrahman committed:
Refactor function to be private.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/186">17.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Add data syncing and pull-to-refresh functionality to the created custom range tab.

hafizrahman committed:
Move function to existing extension.

hafizrahman committed:
Turn timeRangeCount into computed property so that custom range tab is taken into account for syncing after tab change.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -26,16 +28,20 @@ final class StoreStatsAndTopPerformersViewController: TabbedViewController {
         return periodVCs[selection]
     }
 
+    var timeRangeCount: Int {
```
nit: This can be `private` as I don't see ourselves accessing this property from outside this class. 

I would prefer removing this and directly using `timeRanges.count`. In future, if there is a need, we can make `timeRanges` a `private(set)` value. 

What do you think?

hafizrahman replied:
You're right. I overcomplicated that because I misunderstood how Combine works. Updated in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/3594b95927e16fbd52889283e6c56665bcf72e88

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -150,6 +138,23 @@ extension StoreStatsAndTopPerformersViewController: DashboardUI {
 // MARK: - Syncing Data
 //
 private extension StoreStatsAndTopPerformersViewController {
+    func observeSelectedTimeRangeIndex() {
```
Nice work making this private. 👍

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +440,47 @@ private extension StoreStatsAndTopPerformersViewController {
+        let customRangeVC = StoreStatsAndTopPerformersPeriodViewController(siteID: siteID,
+                                                                           timeRange: .custom(from: start, to: end),
+                                                                           currentDate: currentDate,
+                                                                           canDisplayInAppFeedbackCard: true,
```
I think this should be `false` as I see that for other tabs we set this as true only when `timeRange == .today`.

hafizrahman replied:
Nice observation. Handled in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/bac4c2e06f2bcbc028aefea59a8008d8395a50b2

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +440,47 @@ private extension StoreStatsAndTopPerformersViewController {
+
+        let customRangeTabbedItem = TabbedItem(title: Localization.customRangeTabTitle,
+                                               viewController: customRangeVC,
+                                               accessibilityIdentifier: Constants.customRangeTabAcessibilityIdentifier)
```
❓ Can we use the existing `"period-data-" + timeRange.rawValue + "-tab"` pattern to create `accessibilityIdentifier`? 

I am thinking that this will be useful when we save the custom range configuration to disk and recreate the custom range tab in `init`. What do you think?  

One more nice to have: It would be great if we could move initializing `TabbedItem` into a helper method and reuse that from `init` and `createCustomRangeTab`.

hafizrahman replied:
identifier creation updated in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/2a26d87ffaff3240b2040af32a29ed3b545f76fc

> One more nice to have: It would be great if we could move initializing TabbedItem into a helper method and reuse that from init and createCustomRangeTab.

At first I tried this, but ran into a problem: The `TabbedItem` creation in `init` is done inside a closure. If I create a helper method inside `StoreStatsAndTopPerformersViewController` to create `TabbedItem` and call it, it creates this issue:

> 'self' captured by a closure before all members were initialized

The problem is that `periodVCs` is not initialized at that point, so the helper function can't be called yet.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +440,47 @@ private extension StoreStatsAndTopPerformersViewController {
+        removeCustomViewFromTabBar()
+
+        // Get stats data for this tab
+        self.syncStats(forced: false, viewControllerToSync: customRangeVC)
```
nit: `self` can be removed.

hafizrahman replied:
This is included in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/bcd123be653bc29aa63bbe9507d339f8eea68a90

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +440,47 @@ private extension StoreStatsAndTopPerformersViewController {
+                                               accessibilityIdentifier: Constants.customRangeTabAcessibilityIdentifier)
+        appendToTabBar(customRangeTabbedItem)
+
+        // Once a custom range tab is created, do not show the "Custom Range" button anymore.
```
Thanks for adding a comment. 👍

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -431,6 +440,47 @@ private extension StoreStatsAndTopPerformersViewController {
+        periodVCs.append(customRangeVC)
+        timeRanges.append(.custom(from: start, to: end))
+
+        let customRangeTabbedItem = TabbedItem(title: Localization.customRangeTabTitle,
```
We should declare the custom range into a local `let range` and use `range.tabTitle` here. 

`StatsTimeRangeV4` already provides the tab title for the custom range. `Localization.customRangeTabTitle` should be removed.

hafizrahman replied:
TIL I missed that it already has `tabTitle`. Handled in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/bcd123be653bc29aa63bbe9507d339f8eea68a90

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:
```diff
@@ -505,5 +555,14 @@ private extension StoreStatsAndTopPerformersViewController {
 
     enum Constants {
         static let backgroundColor: UIColor = .systemBackground
+        static let customRangeTabAcessibilityIdentifier: String = "period-data-custom-tab"
+    }
+
+    enum Localization {
+        static let customRangeTabTitle = NSLocalizedString(
+            "storeStatsAndTopPerformersViewController.customRangeTabTitle",
+            value: "Custom Range",
+            comment: "Title of the Custom Range tab"
+        )
```
These can be removed, as explained in previous comments.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/bcd123be653bc29aa63bbe9507d339f8eea68a90

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/FilterTabBar/TabbedViewController.swift:
```diff
@@ -102,8 +102,23 @@ class TabbedViewController: UIViewController {
+        tabBar.items = items
+        addChild(tab.viewController)
+        stackView.addArrangedSubview(tab.viewController.view)
+        tab.viewController.didMove(toParent: self)
```
❓  Should we set the `tab.viewController.view` as hidden following `configureChildViewControllers` implementation before we update visibility in `updateVisibleChildViewController`?

hafizrahman replied:
At first I thought we can save some computation by not hiding the custom tab, as it's going to become the selected tab anyway.

But now that I think of it, it might cause some unexpected bug if at some point there's multiple non-hidden view controllers. Added the initial hiding in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/046ac0b3df333cbbf1a19ae46a875408424498e3

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/RangedDatePicker.swift:
```diff
@@ -30,16 +59,22 @@ struct RangedDatePicker: View {
     init(startDate: Date? = nil,
          endDate: Date? = nil,
          datesFormatter: RangedDateTextFormatter,
+         applyButtonTitle: String = Localization.apply,
```
Instead of duplicating the `Localization.apply` string in `RangedDatePickerHostingController`, we can make this parameter optional with `nil` as the default value. 

Inside the `init`, if `applyButtonTitle` is not sent from outside, we can fallback to using `Localization.apply`.

What do you think?

hafizrahman replied:
Great idea, I also renamed the parameter name for better clarity. Done in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/ef8a84cf51dc5bb63baa83248466600c093547e3

selanthiraiyan reviewed (commented):
Thanks for the updates, Hafiz! The custom range loads great. 👏 

I left a few comments on the code. Kindly let me know your thoughts.

hafizrahman committed:
Set this value as false for custom range tab, since only .today time range has this as true.

hafizrahman committed:
Hide custom tab's view controller when it's first created.

It will be shown a little later using the updateVisibleChildViewController() method.

hafizrahman committed:
Refactor range title usage as it is already provided.

hafizrahman committed:
Simplify logic for date picker's custom text for apply button.

hafizrahman committed:
Simplify logic for timeRanges.count checking.

hafizrahman committed:
Simplify accessibility identifier creation.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/RangedDatePicker.swift:
```diff
@@ -7,6 +7,35 @@ protocol RangedDateTextFormatter {
+    }
+}
+
+// MARK: Constant
+
+private extension RangedDatePickerHostingController {
+    enum Localization {
+        static let apply = NSLocalizedString("Apply", comment: "Apply navigation button in custom range date picker")
+    }
+}
```
We should delete this as we don't use this anymore.

hafizrahman replied:
Good catch, thank you. Done in https://github.com/woocommerce/woocommerce-ios/pull/11977/commits/ca820bf0d3a7a47fa12390d4be9271bb97ff5a14

selanthiraiyan reviewed (approved):
Thanks for all the updates, Hafiz! 

🚀 

We can merge this after removing the unused `"Apply"` string.

hafizrahman committed:
Remove unused enum.
